routes/cityListings.js

```diff
@@ -851,6 +855,89 @@ router.post(
     }
 );
 
+router.post(
+    "/:id/vote",
+    async function (req, res, next) {
+        const listingId = req.params.id;
+        const cityId = req.cityId;
+        if (!cityId) {
+            return next(new AppError(`City is not present`, 404));
+        } else {
+            try {
+                const response = await database.get(tables.CITIES_TABLE, {
+                    id: cityId,
+                });
+                if (response.rows && response.rows.length === 0) {
+                    return next(new AppError(`City '${cityId}' not found`, 404));
+                }
+            } catch (err) {
+                return next(new AppError(err));
+            }
+        }
+
+        if (isNaN(Number(listingId)) || Number(listingId) <= 0) {
+            next(new AppError(`Invalid ListingsId ${listingId} given`, 400));
+            return;
+        }
+
+        const optionIdStr = req.body.optionId;
+        if (!optionIdStr) {
+            return next(new AppError(`OptionId not provided`, 400));
+        }
+        const optionId = Number(optionIdStr);
+        if (isNaN(optionId) || optionId <= 0) {
+            return next(new AppError(`Invalid OptionId ${optionId} given`, 400));
+        }
+
+        const requestVote = req.body.vote;
+        if (!requestVote) {
+            return next(new AppError(`Vote not provided`, 400));
+        }
+        const vote = requestVote === "up" ? 1 : -1;
+
+        const response = await database.get(
+            tables.LISTINGS_TABLE,
+            { id: listingId },
+            null,
+            cityId
+        );
+        if (!response.rows || response.rows.length === 0) {
+            return next(
+                new AppError(`Listing with id ${listingId} does not exist`, 404)
+            );
+        }
+        const currentListingData = response.rows[0];
+        if (currentListingData.categoryId !== categories.Polls) {
+            return next(new AppError(`This listing is not a poll`, 400));
+        }
```

Comment:
does it not make more sense to validate listing the above polls and options so it will be easier to debug if needed

Reply:
I do not know for sure. is there any convention for this?